core/src/main/java/com/linecorp/armeria/client/endpoint/healthcheck/PartialHealthCheckStrategy.java

```diff
@@ -0,0 +1,196 @@
+            final Set<Endpoint> removedEndpoints = Sets.difference(selectedEndpoints, this.candidates);
+            removeAndSelectNewEndpoints(removedEndpoints);
+        }
+    }
+
+    @Override
+    public List<Endpoint> getCandidates() {
+        synchronized (selectedEndpoints) {
+            return ImmutableList.copyOf(selectedEndpoints);
+        }
+    }
+
+    @Override
+    public boolean updateHealth(Endpoint endpoint, double health) {
+        final double unhealthyScore = 0;
+
+        requireNonNull(endpoint, "endpoint");
+
+        if (!candidates.contains(endpoint)) {
+            unhealthyEndpoints.remove(endpoint);
+            return true;
+        }
+
+        if (health > unhealthyScore) {
+            unhealthyEndpoints.remove(endpoint);
+            return false;
+        }
+
+        unhealthyEndpoints.add(endpoint);
+        synchronized (selectedEndpoints) {
+            return removeAndSelectNewEndpoints(ImmutableSet.of(endpoint));
+        }
+    }
+
+    /**
+     * This method must be called with synchronized selectedEndpoints.
+     */
+    private boolean removeAndSelectNewEndpoints(Set<Endpoint> removedEndpoints) {
+        final Set<Endpoint> oldSelectedEndpoints = ImmutableSet.copyOf(selectedEndpoints);
+        final int targetSelectedEndpointsSize = max.calculate(candidates.size());
+
+        selectedEndpoints.removeAll(removedEndpoints);
+
+        int availableEndpointsCount = calculateAvailableEndpointsCount(targetSelectedEndpointsSize);
+        if (availableEndpointsCount <= 0) {
+            return true;
+        }
+
+        final int newSelectedEndpointsCount = addRandomlySelectedEndpoints(selectedEndpoints, candidates,
+                                                                           availableEndpointsCount,
+                                                                           selectedEndpoints,
+                                                                           unhealthyEndpoints);
+
+        availableEndpointsCount -= newSelectedEndpointsCount;
+        if (availableEndpointsCount <= 0) {
+            return true;
+        }
+
+        addRandomlySelectedEndpoints(selectedEndpoints, unhealthyEndpoints, availableEndpointsCount,
+                                     selectedEndpoints);
+
+        return !oldSelectedEndpoints.equals(selectedEndpoints);
+    }
+
+    @SafeVarargs
+    private static int addRandomlySelectedEndpoints(Set<Endpoint> selectedEndpoints,
+                                                    Set<Endpoint> candidates, int count,
+                                                    Set<Endpoint>... exclusions) {
+        final List<Endpoint> availableCandidates = new ArrayList<>(candidates.size());
+        loop:
+        for (Endpoint candidate : candidates) {
+            for (Set<Endpoint> exclusion : exclusions) {
+                if (exclusion.contains(candidate)) {
+                    continue loop;
+                }
+            }
```

Comment:
How about remove `varargs` and `loop:` and simplify filter logic?
```java
private static int addRandomlySelectedEndpoints(Set<Endpoint> selectedEndpoints,
                                                Set<Endpoint> candidates, int count,
                                                Set<Endpoint> exclusions) {

    final List<Endpoint> availableCandidates = candidates
            .stream()
            .filter(endpoint -> !exclusions.contains(endpoint))
            .collect(toImmutableList());
    ...

}
```

We could call this method with [`Sets.union(...)`](https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/Sets.java#L725) which does not copy the original sets.

```java
addRandomlySelectedEndpoints(selectedEndpoints, candidates, availableEndpointsCount,
                             Sets.union(selectedEndpoints, unhealthyEndpoints));
```